### PR TITLE
Fix gstreamer backend with manual pipelines

### PR DESCRIPTION
**Pull request by kecsap**
- Fix broken seeking in audio/video playback
- Fix broken audio playback
- Fix unreliable seeking
- Estimate frame count if it is not available directly
- Return -1 for frame count and fps if it is not available. 
- Return 0 for fps if the video has variable frame rate
- Enable and fix tests

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [X] I agree to contribute to the project under Apache 2 License.
- [X] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [X] The PR is proposed to the proper branch
- [-] There is a reference to the original bug report and related work => Reproducible test provided
- [-] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [X] The feature is well documented and sample code can be built with the project CMake

1. Download two test videos:

```bash
wget https://github.com/ietf-wg-cellar/matroska-test-files/raw/master/test_files/test1.mkv
wget https://test-videos.co.uk/vids/jellyfish/mkv/360/Jellyfish_360_10s_5MB.mkv
```

2. I modified a OpenCV videoio sample to demonstrate the problem, here it is the patch: http://dpaste.com//C9MAT2K6W

3. Build the sample, on Ubuntu:

```bash
g++ -g videocapture_audio_combination.cpp -I/usr/include/opencv4 `pkg-config --libs --cflags opencv4` -o videocapture_audio_combination
```

4. Play an audio stream with seeking BEFORE the fix:

```bash
$ ./videocapture_audio_combination --audio "filesrc location=test1.mkv ! queue ! matroskademux name=demux demux.audio_0 ! decodebin ! audioconvert ! appsink"[ERROR:0@0.009] global cap.cpp:164 open VIDEOIO(GSTREAMER): raised OpenCV exception:

OpenCV(4.8.0-dev) ./modules/videoio/src/cap_gstreamer.cpp:153: error: (-215:Assertion failed) ptr in function 'get'


[ WARN:0@0.009] global cap.cpp:204 open VIDEOIO(GSTREAMER): backend is generally available but can't be used to capture by name
ERROR! Can't to open file: filesrc location=test1.mkv ! queue ! matroskademux name=demux demux.audio_0 ! decodebin ! audioconvert ! appsink
```

5. Play a video stream with seeking BEFORE the fix:

```bash
$ ./videocapture_audio_combination --audio "filesrc location=Jellyfish_360_10s_5MB.mkv ! queue ! matroskademux name=demux demux.video_0 ! decodebin ! videoconvert ! video/x-raw, format=BGR ! appsink drop=1"
[ WARN:0@0.034] global cap_gstreamer.cpp:1728 open OpenCV | GStreamer warning: Cannot query video position: status=1, value=22, duration=300
CAP_PROP_AUDIO_DATA_DEPTH: CV_16S
CAP_PROP_AUDIO_SAMPLES_PER_SECOND: 44100
CAP_PROP_AUDIO_TOTAL_CHANNELS: 0
CAP_PROP_AUDIO_TOTAL_STREAMS: [ WARN:0@0.034] global cap_gstreamer.cpp:1898 getProperty OpenCV | GStreamer: CAP_PROP_AUDIO_TOTAL_STREAMS property is not supported
0
[ WARN:0@0.034] global cap_gstreamer.cpp:1817 getProperty OpenCV | GStreamer: CAP_PROP_POS_MSEC property result may be unrealiable: https://github.com/opencv/opencv/issues/19025
Timestamp: 0.6218
Timestamp: 33.1085
Timestamp: 67.1274
Timestamp: 100.1182
Timestamp: 133.1204
Timestamp: 167.1195
Timestamp: 200.1161
Timestamp: 233.1147
Timestamp: 267.1194
Timestamp: 300.1202
[ WARN:0@0.338] global cap_gstreamer.cpp:1949 setProperty OpenCV | GStreamer warning: GStreamer: unable to seek
0:00:00.338215907 3892572 0x5592899c7580 WARN                 basesrc gstbasesrc.c:3127:gst_base_src_loop:<filesrc0> error: Internal data stream error.
0:00:00.338235884 3892572 0x5592899c7580 WARN                 basesrc gstbasesrc.c:3127:gst_base_src_loop:<filesrc0> error: streaming stopped, reason not-linked (-1)
0:00:00.338264287 3892572 0x5592899c7580 WARN                   queue gstqueue.c:992:gst_queue_handle_sink_event:<queue0> error: Internal data stream error.
0:00:00.338270329 3892572 0x5592899c7580 WARN                   queue gstqueue.c:992:gst_queue_handle_sink_event:<queue0> error: streaming stopped, reason not-linked (-1)
[ WARN:0@0.339] global cap_gstreamer.cpp:2784 handleMessage OpenCV | GStreamer warning: Embedded video playback halted; module filesrc0 reported: Internal data stream error.
[ WARN:0@0.339] global cap_gstreamer.cpp:1199 startPipeline OpenCV | GStreamer warning: unable to start pipeline
Number of audio samples: 0
Number of video frames: 10
[ WARN:0@0.339] global cap_gstreamer.cpp:1164 isPipelinePlaying OpenCV | GStreamer warning: GStreamer: pipeline have not been created
```

6. Play an audio stream with seeking AFTER the fix:

```bash
$ ./videocapture_audio_combination --audio "filesrc location=test1.mkv ! queue ! matroskademux name=demux demux.audio_0 ! decodebin ! audioconvert ! appsink"CAP_PROP_AUDIO_DATA_DEPTH: CV_16S
CAP_PROP_AUDIO_SAMPLES_PER_SECOND: 48000
CAP_PROP_AUDIO_TOTAL_CHANNELS: 2
CAP_PROP_AUDIO_TOTAL_STREAMS: [ WARN:0@0.025] global cap_gstreamer.cpp:1903 getProperty OpenCV | GStreamer: CAP_PROP_AUDIO_TOTAL_STREAMS property is not supported
0
Timestamp: 0.0000
Timestamp: 24.0000
Timestamp: 48.0000
Timestamp: 72.0000
Timestamp: 96.0000
Timestamp: 120.0000
Timestamp: 144.0000
Timestamp: 168.0000
Timestamp: 192.0000
Timestamp: 216.0000
Timestamp: 3500.0000
Timestamp: 3504.0000
Timestamp: 3528.0000
Timestamp: 3552.0000
Timestamp: 3576.0000
Timestamp: 3600.0000
Timestamp: 3624.0000
Timestamp: 3648.0000
Timestamp: 3672.0000
Timestamp: 3696.0000
Timestamp: 3720.0000
Timestamp: 3744.0000
Timestamp: 3768.0000
Timestamp: 3792.0000
Timestamp: 3816.0000
Timestamp: 3840.0000
Timestamp: 3864.0000
Timestamp: 3888.0000
Timestamp: 3912.0000
Timestamp: 3936.0000
```

7. Play a video stream with seeking AFTER the fix:

```bash
$ ./videocapture_audio_combination --audio "filesrc location=Jellyfish_360_10s_5MB.mkv ! queue ! matroskademux name=demux demux.video_0 ! decodebin ! videoconvert ! video/x-raw, format=BGR ! appsink drop=1"
[ WARN:0@0.033] global cap_gstreamer.cpp:1746 open OpenCV | GStreamer warning: Cannot query video position: status=1, value=22, duration=300
CAP_PROP_AUDIO_DATA_DEPTH: CV_16S
CAP_PROP_AUDIO_SAMPLES_PER_SECOND: 44100
CAP_PROP_AUDIO_TOTAL_CHANNELS: 0
CAP_PROP_AUDIO_TOTAL_STREAMS: [ WARN:0@0.034] global cap_gstreamer.cpp:1903 getProperty OpenCV | GStreamer: CAP_PROP_AUDIO_TOTAL_STREAMS property is not supported
0
Timestamp: 0.0000
Timestamp: 33.0000
Timestamp: 67.0000
Timestamp: 100.0000
Timestamp: 133.0000
Timestamp: 167.0000
Timestamp: 200.0000
Timestamp: 233.0000
Timestamp: 267.0000
Timestamp: 300.0000
0:00:00.335931693 3893501 0x55bbe76ad920 WARN      matroskareadcommon matroska-read-common.c:759:gst_matroska_read_common_parse_skip:<demux:sink> Unknown CueTrackPositions subelement 0xf0 - ignoring
0:00:00.335952823 3893501 0x55bbe76ad920 WARN      matroskareadcommon matroska-read-common.c:759:gst_matroska_read_common_parse_skip:<demux:sink> Unknown CueTrackPositions subelement 0xf0 - ignoring
0:00:00.335988029 3893501 0x55bbe76ad920 WARN                 basesrc gstbasesrc.c:1742:gst_base_src_perform_seek:<filesrc0> duplicate event found 184
Timestamp: 3467.0000
Timestamp: 3500.0000
Timestamp: 3533.0000
Timestamp: 3567.0000
Timestamp: 3600.0000
Timestamp: 3633.0000
Timestamp: 3667.0000
Timestamp: 3700.0000
Timestamp: 3733.0000
Timestamp: 3767.0000
Timestamp: 3800.0000
Timestamp: 3833.0000
Timestamp: 3867.0000
Timestamp: 3900.0000
Timestamp: 3933.0000
Timestamp: 3967.0000
Timestamp: 4000.0000
Timestamp: 4033.0000
Timestamp: 4067.0000
Timestamp: 4100.0000
```

---

```
force_builders=linux,Linux OpenCL,Linux AVX2,docs
```

### CODE REVIEW

**Comment by kecsap** [comment]
> @kecsap Thanks a lot for your contribution. The GSTreamer issue is known and the related test case is disable for now. Could you enable it and check on your side: https://github.com/opencv/opencv/blob/757d2963823efd00f92041a59bd26d18c2e76728/modules/videoio/test/test_video_io.cpp#L203C18-L203C18

I enabled the tests and fixed them. Along with these actions, a new video duration estimation is implemented if the exact frame count is unavailable. I also added a "queue" element to the default pipeline because it improves the pipeline behavior for certain formats. Although I have to say the test videos do not seem to me very standard, different video tools on Ubuntu Linux report different numbers and metrics.

File: modules/videoio/test/test_video_io.cpp
Comment (kecsap) [reply modules/videoio/test/test_video_io.cpp:245]: Especially, it looks the correct behavior if we can trust other tools. ffprobe reports for big_buck_bunny.mpg:

```bash
Input #0, mpeg, from 'big_buck_bunny.mpg':
  Duration: 00:00:05.08, start: 0.500000, bitrate: 580 kb/s
  Stream #0:0[0x1e0]: Video: mpeg1video, yuv420p(tv), 672x384 [SAR 1:1 DAR 7:4], 104857 kb/s, 24 fps, 24 tbr, 90k tbn, 24 tbc
````

This video starts at 500 msec.

**Comment by kecsap** [comment]
> @kecsap Thanks for the contribution. Looks like the behaviour hardly depends on GStreamer verison and installed plugins set. Some issues With Ubuntu 18.04:
> 

Do not these tests fail before my changes? I did not touch `CAP_PROP_FRAME_COUNT`, only `CAP_PROP_POS_MSEC` in my changes.

**Comment by asmorkalov** [comment]
Ubuntu 22.04 produce the same failures + Ubuntu 22.04 issues. There is mjpeg regression there: https://gitlab.freedesktop.org/gstreamer/gstreamer/-/issues/1122

**Comment by kecsap** [comment]
> Ubuntu 22.04 produce the same failures + Ubuntu 22.04 issues. There is mjpeg regression there: https://gitlab.freedesktop.org/gstreamer/gstreamer/-/issues/1122

Now the mentioned tests pass.

**Comment by kecsap** [comment]
@asmorkalov Pinging... Are the changes okay now?

File: modules/videoio/src/cap_gstreamer.cpp
Comment (asmorkalov) [inline modules/videoio/src/cap_gstreamer.cpp:1549]: Why do you add queue to the pipeline?

File: modules/videoio/src/cap_gstreamer.cpp
Comment (asmorkalov) [inline modules/videoio/src/cap_gstreamer.cpp:1591]: The same for audio queue?

File: modules/videoio/src/cap_gstreamer.cpp
Comment (asmorkalov) [inline modules/videoio/src/cap_gstreamer.cpp:957]: the timestamp has been updated on grabXXXFrame. Looks like it's duplicate and may be removed.

File: modules/videoio/src/cap_gstreamer.cpp
Comment (kecsap) [reply modules/videoio/src/cap_gstreamer.cpp:1591]: Most of the time, it is not necessary and a placebo, but we encountered some files from time to time that were not processed correctly without this GStreamer element.

File: modules/videoio/src/cap_gstreamer.cpp
Comment (kecsap) [reply modules/videoio/src/cap_gstreamer.cpp]: Done

**Comment by kecsap** [comment]
I addressed the comments and rebased the comment on the top of the branch 4.x.

**Comment by kecsap** [comment]
@mshabunin Ping-pong. Please review if you have time.

**Review by asmorkalov — APPROVED** [review]
:+1:

**Comment by kecsap** [comment]
Are there any other things to do on my side before you merge? Waiting for second approval? Rebasing?

File: modules/videoio/src/cap_gstreamer.cpp
Comment (mshabunin) [inline modules/videoio/src/cap_gstreamer.cpp:1724]: I would be good to have simple RAII wrapper class for this object.

File: modules/videoio/src/cap_gstreamer.cpp
Comment (mshabunin) [inline modules/videoio/src/cap_gstreamer.cpp:1739]: Shouldn't `duration` be set to `-1` in case of error?

File: modules/videoio/src/cap_gstreamer.cpp
Comment (mshabunin) [inline modules/videoio/src/cap_gstreamer.cpp:1736]: What aboud VFR videos? How will it work?

File: modules/videoio/src/cap_gstreamer.cpp
Comment (mshabunin) [inline modules/videoio/src/cap_gstreamer.cpp:1960]: Shouldn't we stop pipeline for POS_MSEC property? Or POS_AVI_RATIO? Or WIDTH/HEIGHT?

File: modules/videoio/src/cap_gstreamer.cpp
Comment (kecsap) [reply modules/videoio/src/cap_gstreamer.cpp]: Now I made it changed and made it clear in the commit message. VFR does not calculate frame count and return 0 for fps. If fps query fails fps equals to -1.

**Comment by mshabunin** [comment]
BTW, I have this example app which works similarly to OpenCV GStreamer capture (https://github.com/mshabunin/gstreamer-example) and it demonstrates that GStreamer might not be reliable when evaluating stream position and timestamps.

E.g. `gst_element_query_position` can return `1300 ms` before first sample being retrieved, or first BUFFER_PTS can be `200 ms` for mp4 file and `0 ms` for mkv file.

File: modules/videoio/src/cap_gstreamer.cpp
Comment (mshabunin) [inline modules/videoio/src/cap_gstreamer.cpp:1860]: Will this work after seeking to different position? As I can see `timstamp` is being updated only during frame reading.

File: modules/videoio/src/cap_gstreamer.cpp
Comment (mshabunin) [inline modules/videoio/src/cap_gstreamer.cpp:1737]: Could you please reformat this block (curly braces) to be consistent with the whole function.

**Review by mshabunin — COMMENTED** [review]
It would be nice to have GStreamer-specific seeking test, because we don't have one (see `test_gstreamer.cpp`). It can use any video file from testdata (preferably the one which is the most reliable) or generate temporary synthetic file and also check both manual and automatic pipeline construction.

File: modules/videoio/src/cap_gstreamer.cpp
Comment (kecsap) [reply modules/videoio/src/cap_gstreamer.cpp]: I checked earlier and stopping-restarting was an old implementation in the sources. When I checked GStreamer docs, I found no reference to why the pipeline should be stopped for any seeking or other behavior.
The latest version stops the pipeline for fps-image width and image height changes. I also found mov and mp4 files where a temporal pipeline stop is needed. It is added with comment.
POS_AVI_RATIO: It is basically not supported in GStreamer. I added a new OpenCV warning and comment to the source code about it.
CV_CAP_PROP_POS_FRAMES: I could not make it work.

File: modules/videoio/src/cap_gstreamer.cpp
Comment (kecsap) [reply modules/videoio/src/cap_gstreamer.cpp]: Done

File: modules/videoio/src/cap_gstreamer.cpp
Comment (kecsap) [reply modules/videoio/src/cap_gstreamer.cpp:1863]: Now it is cached in setProp() event with optimism if it will be successful.

**Review by mshabunin — COMMENTED** [review]
> I politely would like to express, it is not my business, I am not a GStreamer developer. I made changes based on timestamps extracted from the acquired buffers which is much more reliable than the earlier implementation in this backend and it does not crash as before in some cases. It is far from perfect, but much better than it was before.

> "It would be nice to have GStreamer-specific seeking test, because we don't have one (see test_gstreamer.cpp)."

> Is not it already there in modules/videoio/test/test_video_io.cpp?

Apparently some scenarios are covered. You've fixed some scenario (`I modified a OpenCV videoio sample to demonstrate the problem`), but the paste you shared is not available and it is not covered by our tests. Is it possible to add it to the tests to be sure future modifications wouldn't break it?

__Note:__
I've prepared a framework for OpenCV-GStreamer testing (https://github.com/mshabunin/gstreamer-check). It helped to determine following things:
* for some reason GStreamer _audio_ tests hang or crash or fail in both cases: with patch and without patch
* MJPG-related GStreamer tests fail only in Ubuntu 22, in more recent versions of GStreamer (Ubuntu 23, Fedora, manually-built latest GStreamer), these tests pass

This behavior can be related to the system configuration or installed plugins or HW availability in container, because on my local system all tests pass with manually-built latest GStreamer release.

File: modules/videoio/src/cap_gstreamer.cpp
Comment (kecsap) [reply modules/videoio/src/cap_gstreamer.cpp]: Done

File: modules/videoio/src/cap_gstreamer.cpp
Comment (kecsap) [reply modules/videoio/src/cap_gstreamer.cpp]: Done

File: modules/videoio/src/cap_gstreamer.cpp
Comment (kecsap) [reply modules/videoio/src/cap_gstreamer.cpp]: I applied `needRestart`.
If you check other properties, all return in the switch statement except those I check by `propId`. It does not make any sense to stop the pipeline for the other. Especially it was a bug to stop the pipeline and return for other properties.

File: modules/videoio/src/cap_gstreamer.cpp
Comment (kecsap) [reply modules/videoio/src/cap_gstreamer.cpp:2027]: It is not necessary. When I test with those files, the pipeline works without explicit restart. And you can check the old source version and my other comment: CV_CAP_PROP_POS_FRAMES branch returned without restart before my changes.

**Comment by kecsap** [comment]
@mshabunin 

Okay, a new round of updates are up:
- I fixed your comments.
- I enabled back the MJPG test cases.
- I complemented a test with a manual GStreamer pipeline seeking case in `test_video_io.cpp`.

**Review by mshabunin — APPROVED** [review]
I've moved new test to test_gstreamer.cpp and tested it on various platforms.

**Pull request merged**
